auto_run.py

```diff
@@ -0,0 +1,146 @@
+            shuffle=False,  
+            batch_size=len(test_df),
+            online_fetch=False)
+
+    # Select image to predict heatmaps
+    X_batch, y_stacked = generator[0] # There is only one batch in the generator
+    # X_batch, y_stacked = evaluation.load_and_preprocess_img('data/skier.jpg', eval.num_hg_blocks)
+    y_batch = y_stacked[0] # take first hourglass section
+    # Save stacked heatmap images to disk
+    m_batch = test_df.to_dict('records') # TODO: eventually this will be passed from data generator as metadata
+    print("\n\nEval start:   {}\n".format(time.ctime()))
+    eval.visualize_batch(X_batch, y_batch, m_batch)
+    print("\n\nEval end:   {}\n".format(time.ctime()))
+
+#%% 
+
+def predict_kp_bbox(val_df, eval):
+    generator = data_generator.DataGenerator(
+                df=val_df,
+                base_dir=DEFAULT_VAL_IMG_PATH,
+                input_dim=INPUT_DIM,
+                output_dim=OUTPUT_DIM,
+                num_hg_blocks=DEFAULT_NUM_HG,
+                shuffle=False,  
+                batch_size=1,
+                online_fetch=False)
+
+    # Select image to predict heatmaps
+    X_batch, y_stacked = generator[168] # choose one image for evaluation
+    y_batch = y_stacked[0] # take first hourglass section
+    X, y = X_batch[0], y_batch[0] # take first example of batch
+
+    # Get predicted heatmaps for image
+    predict_heatmaps=eval.predict_heatmaps(X_batch)
+
+    # Get predicted keypoints from last hourglass (eval.num_hg_blocks-1)
+    keypoints = eval.heatmaps_to_keypoints(predict_heatmaps[eval.num_hg_blocks-1, 0, :, :, :])
+    print(keypoints)
+    # Get bounding box image from heatmap
+    heatmap = y[:,:,0]
+    hm = HeatMap(X,heatmap)
+    img = np.array(hm.image)
+
+    # Clear plot image
+    plt.clf()
+    # Plot predicted keypoints on bounding box image
+    x = []
+    y = []
+    for i in range(NUM_COCO_KEYPOINTS):
+        if(keypoints[i,0] != 0 and keypoints[i,1] != 0):
+            x.append(keypoints[i,0])
+            y.append(keypoints[i,1])
+    plt.scatter(x,y)
+    plt.imshow(img)
+
+
+# %% find each model's cooresponding epochs 
+
+h = HourglassNet(NUM_COCO_KEYPOINTS,DEFAULT_NUM_HG,INPUT_CHANNELS,INPUT_DIM,OUTPUT_DIM)
+_, val_df = h.load_and_filter_annotations(DEFAULT_TRAIN_ANNOT_PATH,DEFAULT_VAL_ANNOT_PATH,0.1)
+
+representative_set_df = pd.read_pickle(os.path.join(DEFAULT_PICKLE_PATH, 'representative_set.pkl'))
+models = [] # keep track of the models that are visited 
+epoch = [] # extract out all epoch numbers (across different files) from model logs
+output_models = os.listdir(DEFAULT_OUTPUT_BASE_DIR) # keep track of the models have been explored before
+
+for sub_dir in os.listdir(DEFAULT_MODEL_BASE_DIR):
+    if '_hg_' in sub_dir and sub_dir not in models and sub_dir not in output_models:
+        models.append(sub_dir)
+        epoch = find_epochs(DEFAULT_MODEL_BASE_DIR, sub_dir, epoch, models)
+        for (n_epoch, model_file) in epoch:
```

Comment:
Since the models make large progress at the begging and less so near the end, and because loading/visualizing takes some time, we may want to limit the epochs we visualize. So for example, only visualize epochs: `[1,2,3,4,5,10,20,30,40]` or something along those lines

Reply:
That is a very good point! I'll look into it :)